### PR TITLE
Support for SIZE command and cross-compilation on Linux to Windows

### PR DESCRIPTION
**Pull request by bjuulp**
This pull request adds support for the SIZE command. Preliminary support for cross-compilation on Linux to Windows is also added. See the example CMake toolchain file. There are some issues wrt support for wstring in the fstream constructor when cross-compiling with GCC 7.5.0. I've therefore had to disable the UTF8 supporting code in 1 place. The changes have been tested on Windows (based a cross-compiled exe built on Linux), only.

### CODE REVIEW

File: fineftp-server/src/ftp_session.cpp [hunk 1]
```diff
@@ -485,6 +486,48 @@ namespace fineftp
...
+    {
+      sendFtpMessage(FtpReplyCode::ACTION_NOT_TAKEN, "Permission denied");
+      return;
+    }
+
+    std::string local_path = toLocalPath(param);
+
+    std::ios::openmode open_mode =
+       std::ios::ate | (data_type_binary_ ? (std::ios::in | std::ios::binary) : (std::ios::in));
+    std::ifstream file(local_path, open_mode);
```
Comment (FlorianReimold) [inline fineftp-server/src/ftp_session.cpp:509]: The local_path must convert to an UTF16 wstring on windows

File: fineftp-server/src/filesystem.cpp [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 #define NOMINMAX
 #define WIN32_LEAN_AND_MEAN
-#include <Windows.h>
+#include <windows.h>
```
Comment (FlorianReimold) [inline fineftp-server/src/filesystem.cpp:12]: Is there a reason for this edit? At least on my machine the file is:
`C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\Windows.h`
So with a capital `W`.

**Review by FlorianReimold — CHANGES_REQUESTED** [review]
Great work, thank you! 👍
I basically only have one request for a change and that is to add unicode support in the SIZE call on Windows. This means that one will have to convert the UTF8 encoded std::string to a UTF16 encoded std::wstring, as otherwise the outcome will depend on the Codepage (and we live in 2022 now and don't want to rely on Windows Codepages 😄)

File: fineftp-server/src/filesystem.cpp (hunk 2 above)
Comment (bjuulp) [reply fineftp-server/src/filesystem.cpp:12]: When cross-compiling on Linux for Windows, this file is spelled with lower case w. I believe the compiler on Windows is insensitive to the case of the filename, so the change should be working on Windows, too, although I haven't had the possibility to try that.

File: fineftp-server/src/ftp_session.cpp (hunk 1 above)
Comment (bjuulp) [reply fineftp-server/src/ftp_session.cpp:509]: Thanks. I will look into that.

**Comment by FlorianReimold** [comment]
I have merged the changes to a branch and will now check whether I can resolve the mingw issue that you wrote about. mingw is a compiler that we try to support at least a little bit, so it would be nice if UTF8 support works for that.

**Comment by bjuulp** [comment]
Thanks. I'm not sure the current use of std::ftream can deal with this. I think we will have to go directly on the Win32 API to solve this for both mingw and msvc. But that may be needed for another reason, too: I'd like to see memory mapped files being used instead in order to increase speed and reduce pressure on memory. I have a working example in some other FTP server code I have created. That could potentially refactored to fit into the fineftp-server as well, although it will take some time for me to do that. I'm rather busy now, but will see if I can get that done within a month or so.

**Comment by FlorianReimold** [comment]
I am not sure either, but I will try anyways. If you have already running code for memory mapped files I will be happe to accept another PR 😊

**Pull request merged**